### PR TITLE
New ledger transactions

### PR DESCRIPTION
**Pull request by snopoke**
@czue 

@dimagi/scale-team 

Still outstanding (probably for another PR):
- [ ] move consumption calculation to async process
- [ ] use consumption from StockState / LedgerValue in restore instead of re-calculating it (https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/data_providers/case/stock.py#L26-L26)

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 46 - E501 line too long (124 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 55 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 156 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 162 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 168 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 175 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 210 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 214 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 227 - E121 continuation line under-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 256 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 258 - E226 missing whitespace around arithmetic operator
* corehq/form_processor/backends/sql/dbaccessors.py, line 15 - E501 line too long (116 > 115 characters)
* corehq/form_processor/backends/sql/ledger.py, line 50 - E501 line too long (119 > 115 characters)
* corehq/form_processor/models.py, line 1062 - E203 whitespace before ','
* corehq/form_processor/tests/test_ledgers.py, line 156 - E123 closing bracket does not match indentation of opening bracket's line

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 46 - E501 line too long (124 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 55 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 156 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 162 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 168 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 175 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 210 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 214 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 227 - E121 continuation line under-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 256 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 258 - E226 missing whitespace around arithmetic operator
* corehq/form_processor/backends/sql/dbaccessors.py, line 15 - E501 line too long (116 > 115 characters)
* corehq/form_processor/backends/sql/ledger.py, line 50 - E501 line too long (119 > 115 characters)
* corehq/form_processor/models.py, line 1062 - E203 whitespace before ','
* corehq/form_processor/tests/test_ledgers.py, line 156 - E123 closing bracket does not match indentation of opening bracket's line

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 46 - E501 line too long (124 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 55 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 156 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 162 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 168 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 175 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 210 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 214 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 227 - E121 continuation line under-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 256 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 258 - E226 missing whitespace around arithmetic operator
* corehq/form_processor/backends/sql/dbaccessors.py, line 15 - E501 line too long (116 > 115 characters)
* corehq/form_processor/backends/sql/ledger.py, line 50 - E501 line too long (119 > 115 characters)
* corehq/form_processor/models.py, line 1062 - E203 whitespace before ','
* corehq/form_processor/tests/test_ledgers.py, line 156 - E123 closing bracket does not match indentation of opening bracket's line

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 46 - E501 line too long (124 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 55 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 156 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 162 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 168 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 175 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 210 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 214 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 227 - E121 continuation line under-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 256 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 258 - E226 missing whitespace around arithmetic operator
* corehq/form_processor/backends/sql/dbaccessors.py, line 15 - E501 line too long (116 > 115 characters)
* corehq/form_processor/backends/sql/ledger.py, line 50 - E501 line too long (124 > 115 characters)
* corehq/form_processor/models.py, line 1062 - E203 whitespace before ','
* corehq/form_processor/tests/test_ledgers.py, line 156 - E123 closing bracket does not match indentation of opening bracket's line

**Comment by czue** [comment]
wow, seeing it all together like this makes me realize what a mess this code is. nice effort!

mostly minor/clarifying comments. assume you'll look into build failures.

```
======================================================================
FAIL: test_thresholds (casexml.apps.stock.tests.test_consumption_calc.ConsumptionCalcTestNew)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py", line 255, in test_thresholds
    self.assertEqual(self.consumption(tx, 60, {'min_periods': 4}), None)
AssertionError: 6.0 != None
```

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 46 - E501 line too long (124 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/base.py, line 55 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 160 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 166 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 172 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 179 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 214 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 218 - E126 continuation line over-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 231 - E121 continuation line under-indented for hanging indent
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 260 - E226 missing whitespace around arithmetic operator
* corehq/ex-submodules/casexml/apps/stock/tests/test_consumption_calc.py, line 262 - E226 missing whitespace around arithmetic operator
* corehq/form_processor/backends/sql/dbaccessors.py, line 15 - E501 line too long (116 > 115 characters)
* corehq/form_processor/backends/sql/ledger.py, line 50 - E501 line too long (124 > 115 characters)
* corehq/form_processor/models.py, line 1062 - E203 whitespace before ','
* corehq/form_processor/tests/test_ledgers.py, line 156 - E123 closing bracket does not match indentation of opening bracket's line

File: corehq/form_processor/tests/test_ledgers.py [hunk 1]
```diff
@@ -4,29 +4,19 @@
...
 from corehq.apps.commtrack.helpers import make_product
+from corehq.apps.commtrack.tests import get_single_balance_block
+from corehq.apps.commtrack.tests.util import get_single_transfer_block
 from corehq.apps.hqcase.utils import submit_case_blocks
-from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL
+from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL, LedgerAccessorSQL
 from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
 from corehq.form_processor.interfaces.processor import FormProcessorInterface
-from corehq.form_processor.models import CaseTransaction
+from corehq.form_processor.models import CaseTransaction, LedgerTransaction
```
Comment (dimagimon) [inline corehq/form_processor/tests/test_ledgers.py]: F401 'CaseTransaction' imported but unused

File: corehq/form_processor/tests/test_ledgers.py [hunk 2]
```diff
@@ -74,11 +72,15 @@ def test_balance_submission_multiple(self):
             self.product_c._id: 25,
         }
         self._submit_ledgers([
-            TestLedger(BALANCE_BLOCK, balance, prod_id)
+            get_single_balance_block(self.case.case_id, prod_id, balance)
```
Comment (dimagimon) [inline corehq/form_processor/tests/test_ledgers.py]: F821 undefined name 'get_single_balance_block'

File: corehq/form_processor/tests/test_ledgers.py [hunk 3]
```diff
@@ -95,29 +99,61 @@ def test_balance_submission_with_prior_balance(self):
...
     @run_with_all_backends
     def test_ledger_update_with_case_update(self):
         submit_case_blocks([
             CaseBlock(case_id=self.case.case_id, update={'a': "1"}).as_string(),
-            BALANCE_BLOCK.format(
-                case_id=self.case.case_id,
-                product_id=self.product_a._id,
-                quantity=100
-            )],
+            get_single_balance_block(self.case.case_id, self.product_a._id, 100)],
```
Comment (dimagimon) [inline corehq/form_processor/tests/test_ledgers.py]: F821 undefined name 'get_single_balance_block'

**Comment by benrudolph** [comment]
just skimmed, but it's looking good!

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by czue** [comment]
thanks for responding to feedback

**Comment by czue** [comment]
:dollar: :book:

**Pull request merged**